File: runtime/android/app/src/main/java/com/appunvs/runtime/ui/Bubble.kt
// Bubble — chat message.  User messages right-align with brand fill;
// assistant / system messages left-align with bordered surface.  Soft
// "tail" corner trims the relevant top corner per role.
package com.appunvs.runtime.ui

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp

import com.appunvs.runtime.theme.LocalAppColors
import com.appunvs.runtime.theme.Radius
import com.appunvs.runtime.theme.Spacing

enum class BubbleRole { USER, ASSISTANT, SYSTEM }

@Composable
fun Bubble(
    role: BubbleRole,
    text: String,
    pending: Boolean = false,
    modifier: Modifier = Modifier,
) {
    val colors = LocalAppColors.current
    val isUser = role == BubbleRole.USER

    val bg = when (role) {
        BubbleRole.USER      -> colors.brandDark
        BubbleRole.ASSISTANT -> colors.bgCard
        BubbleRole.SYSTEM    -> colors.bgInput
    }
    val fg = if (isUser) Color.White else colors.textPrimary

    val shape = if (isUser) {
        RoundedCornerShape(
            topStart = Radius.xl.dp,
            bottomStart = Radius.xl.dp,
            bottomEnd = Radius.xl.dp,
            topEnd = Radius.s.dp,
        )
    } else {
        RoundedCornerShape(
            topStart = Radius.s.dp,
            bottomStart = Radius.xl.dp,
            bottomEnd = Radius.xl.dp,
            topEnd = Radius.xl.dp,
        )
    }

    val display = if (text.isEmpty() && pending) "…" else text

    Row(
        modifier = modifier.fillMaxWidth(),
        horizontalArrangement = if (isUser) Arrangement.End else Arrangement.Start,
        verticalAlignment = Alignment.Top,
    ) {
        Text(
            text = display,
            color = fg,
            modifier = Modifier
                .widthIn(max = 360.dp)
                .clip(shape)
                .background(bg)
                .let {
                    if (!isUser) it.border(BorderStroke(1.dp, colors.borderDefault), shape) else it
                }
                .padding(horizontal = Spacing.l.dp, vertical = Spacing.m.dp),
        )
    }
}

File: runtime/ios/Runtime/UI/Bubble.swift
// Bubble — chat message.  User messages right-align with brand fill;
// assistant / system messages left-align with bordered surface.  Max
// width caps at 85% of the parent so two long bubbles can sit side by
// side on the iPad split view.
import SwiftUI

enum BubbleRole {
    case user, assistant, system
}

struct Bubble: View {
    let role: BubbleRole
    let text: String
    let pending: Bool

    init(role: BubbleRole, text: String, pending: Bool = false) {
        self.role = role
        self.text = text
        self.pending = pending
    }

    var body: some View {
        HStack {
            if role == .user { Spacer(minLength: 0) }
            Text(displayText)
                .font(.body)
                .foregroundStyle(foreground)
                .padding(.horizontal, Spacing.l)
                .padding(.vertical, Spacing.m)
                .background(
                    UnevenRoundedRectangle(
                        cornerRadii: corners,
                        style: .continuous
                    )
                    .fill(background)
                )
                .overlay(
                    role == .user
                        ? nil
                        : UnevenRoundedRectangle(
                            cornerRadii: corners,
                            style: .continuous
                        )
                        .stroke(Theme.borderDefault.color, lineWidth: 1)
                )
                .frame(maxWidth: .infinity, alignment: role == .user ? .trailing : .leading)
            if role != .user { Spacer(minLength: 0) }
        }
    }

    private var displayText: String {
        if text.isEmpty && pending { return "…" }
        return text
    }

    private var foreground: Color {
        if role == .user {
            return Color.white
        }
        return Theme.textPrimary.color
    }

    private var background: Color {
        switch role {
        case .user:      return Theme.brandDark.color
        case .assistant: return Theme.bgCard.color
        case .system:    return Theme.bgInput.color
        }
    }

    /// User bubbles get a tighter top-right corner (the "tail" side);
    /// assistant bubbles get a tighter top-left.  The other three
    /// corners stay at radius.xl for the soft "pillow" feel.
    private var corners: RectangleCornerRadii {
        if role == .user {
            return .init(topLeading: Radius.xl, bottomLeading: Radius.xl,
                         bottomTrailing: Radius.xl, topTrailing: Radius.s)
        } else {
            return .init(topLeading: Radius.s, bottomLeading: Radius.xl,
                         bottomTrailing: Radius.xl, topTrailing: Radius.xl)
        }
    }
}

#Preview {
    VStack(alignment: .leading, spacing: 8) {
        Bubble(role: .user, text: "做一个计数器 app")
        Bubble(role: .assistant, text: "好的,我来写。")
        Bubble(role: .assistant, text: "", pending: true)
    }
    .padding()
    .background(Theme.bgPage.color)
}
